File: rwanda-locations-json-master/analyze_locations.js
const fs = require('fs');

function analyzeRwandaLocations() {
    const data = JSON.parse(fs.readFileSync('locations.json', 'utf8'));
    
    // Count all levels
    const provinces = data.provinces.length;
    const districts = data.provinces.reduce((sum, p) => sum + p.districts.length, 0);
    const sectors = data.provinces.reduce((sum, p) => 
        sum + p.districts.reduce((dSum, d) => dSum + d.sectors.length, 0), 0);
    const cells = data.provinces.reduce((sum, p) => 
        sum + p.districts.reduce((dSum, d) => 
            dSum + d.sectors.reduce((sSum, s) => sSum + s.cells.length, 0), 0), 0);
    const villages = data.provinces.reduce((sum, p) => 
        sum + p.districts.reduce((dSum, d) => 
            dSum + d.sectors.reduce((sSum, s) => 
                sSum + s.cells.reduce((cSum, c) => cSum + c.villages.length, 0), 0), 0), 0);
    
    console.log('🇷🇼 RWANDA COMPLETE ADMINISTRATIVE STRUCTURE');
    console.log('='.repeat(50));
    console.log(`📍 Provinces: ${provinces}`);
    console.log(`📍 Districts: ${districts}`);
    console.log(`📍 Sectors: ${sectors}`);
    console.log(`📍 Cells: ${cells}`);
    console.log(`📍 Villages: ${villages}`);
    console.log(`📍 Total Administrative Units: ${villages + cells + sectors + districts + provinces}`);
    
    return { data, stats: { provinces, districts, sectors, cells, villages } };
}

function searchLocation(data, searchTerm) {
    const results = [];
    
    data.provinces.forEach(province => {
        province.districts.forEach(district => {
            district.sectors.forEach(sector => {
                sector.cells.forEach(cell => {
                    cell.villages.forEach(village => {
                        if (village.name.toLowerCase().includes(searchTerm.toLowerCase())) {
                            results.push({
                                village: village.name,
                                cell: cell.name,
                                sector: sector.name,
                                district: district.name,
                                province: province.name,
                                fullPath: `${village.name} → ${cell.name} → ${sector.name} → ${district.name} → ${province.name}`
                            });
                        }
                    });
                });
            });
        });
    });
    
    return results;
}

function getAllLocations(data) {
    const allLocations = {
        provinces: [],
        districts: [],
        sectors: [],
        cells: [],
        villages: []
    };
    
    data.provinces.forEach(province => {
        allLocations.provinces.push(province.name);
        
        province.districts.forEach(district => {
            allLocations.districts.push({
                name: district.name,
                province: province.name
            });
            
            district.sectors.forEach(sector => {
                allLocations.sectors.push({
                    name: sector.name,
                    district: district.name,
                    province: province.name
                });
                
                sector.cells.forEach(cell => {
                    allLocations.cells.push({
                        name: cell.name,
                        sector: sector.name,
                        district: district.name,
                        province: province.name
                    });
                    
                    cell.villages.forEach(village => {
                        allLocations.villages.push({
                            name: village.name,
                            cell: cell.name,
                            sector: sector.name,
                            district: district.name,
                            province: province.name
                        });
                    });
                });
            });
        });
    });
    
    return allLocations;
}

// Export for use in other files
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        analyzeRwandaLocations,
        searchLocation,
        getAllLocations
    };
}

// Run analysis if called directly
if (require.main === module) {
    const { data, stats } = analyzeRwandaLocations();
    
    console.log('\n🔍 Example: Search for "Kigali"');
    const results = searchLocation(data, 'Kigali').slice(0, 5);
    results.forEach(r => console.log(`   ${r.fullPath}`));
    
    console.log(`\n✅ Analysis complete! All ${stats.villages} villages from ${stats.provinces} provinces are available.`);
}

File: rwanda-locations-json-master/analyze_locations.py
import json

def analyze_rwanda_locations():
    with open('locations.json', 'r', encoding='utf-8') as f:
        data = json.load(f)
    
    # Count all administrative levels
    provinces = len(data['provinces'])
    districts = sum(len(p['districts']) for p in data['provinces'])
    sectors = sum(len(d['sectors']) for p in data['provinces'] for d in p['districts'])
    cells = sum(len(s['cells']) for p in data['provinces'] for d in p['districts'] for s in d['sectors'])
    villages = sum(len(c['villages']) for p in data['provinces'] for d in p['districts'] for s in d['sectors'] for c in s['cells'])
    
    print(f"🇷🇼 RWANDA COMPLETE ADMINISTRATIVE STRUCTURE")
    print(f"=" * 50)
    print(f"📍 Provinces: {provinces}")
    print(f"📍 Districts: {districts}")
    print(f"📍 Sectors: {sectors}")
    print(f"📍 Cells: {cells}")
    print(f"📍 Villages: {villages}")
    print(f"📍 Total Lines: {villages + cells + sectors + districts + provinces}")
    
    # List all provinces and their districts
    print(f"\n📋 COMPLETE STRUCTURE:")
    for i, province in enumerate(data['provinces'], 1):
        print(f"\n{i}. PROVINCE: {province['name']}")
        for j, district in enumerate(province['districts'], 1):
            print(f"   {i}.{j} District: {district['name']} ({len(district['sectors'])} sectors)")
    
    return data

def search_location(data, term):
    results = []
    for province in data['provinces']:
        for district in province['districts']:
            for sector in district['sectors']:
                for cell in sector['cells']:
                    for village in cell['villages']:
                        if term.lower() in village['name'].lower():
                            results.append({
                                'village': village['name'],
                                'cell': cell['name'],
                                'sector': sector['name'],
                                'district': district['name'],
                                'province': province['name']
                            })
    return results

def get_province_data(data, province_name):
    for province in data['provinces']:
        if province_name.lower() in province['name'].lower():
            return province
    return None

if __name__ == "__main__":
    data = analyze_rwanda_locations()
    
    # Example searches
    print(f"\n🔍 Search example - Villages with 'Kigali':")
    results = search_location(data, "Kigali")[:3]
    for r in results:
        print(f"   {r['village']} → {r['cell']} → {r['sector']} → {r['district']} → {r['province']}")